### PR TITLE
fix(state-sync): Fix catchup updating flat storage

### PR DESCRIPTION
**Pull request by nikurt**
I observed test nodes getting stuck after catching up.
Fix the issue by making the node always open flat storage for all shards.

The issue is complicated:
* Node tracks a shard.
* Node restarts without tracking a shard.
* Because the shard is not tracked, no more changes to that shard will be reflected in flat storage.
* Perform catchup, which **adds** flat state entities, but doesn't delete pre-existing flat state entities.
* Observe that flat state is no longer matching the trie state, which will eventually cause wrong state roots.

Add a test reproducing the issue.

### CODE REVIEW

File: pytest/tests/sanity/catchup_flat_storage_deletions.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    assert epoch == 4, f"epoch: {epoch}"
+    logger.info(f'We are in epoch {epoch}')
+
+    # Ensure the non-validator node has caught up.
+    utils.wait_for_blocks(node, target=int(4.5 * EPOCH_LENGTH))
+    logger.info(f'The other node is in sync')
+
+    print_balances(nodes, account_ids)
+
+    # Send tokens and expect the transaction to fail.
```
Comment (jakmeier) [inline pytest/tests/sanity/catchup_flat_storage_deletions.py:150]: Is this comment accurate? It seems like we are checking that the transaction was accepted by the RPC, not that executing it failed. At least I can't find any checks on the transaction outcome or actual balances.

File: pytest/tests/sanity/catchup_flat_storage_deletions.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
+#!/usr/bin/env python3
+# Spins up a validator node tracking all shards and a non-validator node.
+# Deletes an account.
+# Restart the non-validator node.
+# After the non-validator node does catchup, attempt to send a token to the deleted account.
+# Observe that both nodes correctly execute the transaction and return an error.
```
Comment (jakmeier) [inline pytest/tests/sanity/catchup_flat_storage_deletions.py:6]: Is this observation verified in any way, or just something you could see in the printed logs?

**Review by jakmeier — COMMENTED** [review]
The fix looks ok to me. I'm not sure if anyone ever runs neard in a setup where tracking specific shards is required. I suspect not, but do you have more insights here?

For the test, it looks to me like the comments describe a bit more rigorous checking than the code. But I guess the test code is sufficient to reproduce the error and check that it was fixed, so I don't really mind.

File: pytest/tests/sanity/catchup_flat_storage_deletions.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    node.kill()
+    # Restart the node to make it start without opening any flat storages.
+    node.start(boot_node=boot_node)
+    logger.info(f'We are in epoch {epoch}, and the node is restarted')
+
+    print_balances(nodes, account_ids)
+
+    # Delete the account.
+    latest_block_hash = boot_node.get_latest_block().hash_bytes
+    nonce += 1
```
[the remark this answers is not in the document]
Comment (nikurt) [reply pytest/tests/sanity/catchup_flat_storage_deletions.py:129]: As the accounts are created in the genesis, the accepted nonces are [0,1M].

File: pytest/tests/sanity/catchup_flat_storage_deletions.py (hunk 1 above)
Comment (nikurt) [reply pytest/tests/sanity/catchup_flat_storage_deletions.py]: adjusted the comment, thanks.

File: pytest/tests/sanity/catchup_flat_storage_deletions.py (hunk 2 above)
Comment (nikurt) [reply pytest/tests/sanity/catchup_flat_storage_deletions.py:6]: It's verified by the second node not getting stuck.
Added a direct check that `test_account_id` doesn't exist on the node.

File: pytest/tests/sanity/catchup_flat_storage_deletions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jakmeier) [reply pytest/tests/sanity/catchup_flat_storage_deletions.py:129]: I see, makes sense  :+1:

**Review by jakmeier — APPROVED** [review]

**Pull request merged**